File: src/main/java/it/frob/sleighidea/SleighParserDefinition.java
// SPDX-License-Identifier: Apache-2.0

package it.frob.sleighidea;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import it.frob.sleighidea.parser.SleighParser;
import it.frob.sleighidea.psi.SleighFile;
import it.frob.sleighidea.psi.SleighTypes;
import org.jetbrains.annotations.NotNull;

public class SleighParserDefinition implements ParserDefinition {

    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(SleighTypes.COMMENT);

    public static final IFileElementType FILE = new IFileElementType(SleighLanguage.INSTANCE);

    @Override
    public @NotNull TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @Override
    public @NotNull SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
        return SpaceRequirements.MAY;
    }

    @Override
    public @NotNull Lexer createLexer(Project project) {
        return new SleighLexerAdapter();
    }

    @Override
    public @NotNull PsiParser createParser(Project project) {
        return new SleighParser();
    }

    @Override
    public @NotNull IFileElementType getFileNodeType() {
        return FILE;
    }

    @Override
    public @NotNull TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @Override
    public @NotNull TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @Override
    public @NotNull PsiElement createElement(ASTNode node) {
        return SleighTypes.Factory.createElement(node);
    }

    @Override
    public @NotNull PsiFile createFile(@NotNull FileViewProvider viewProvider) {
        return new SleighFile(viewProvider);
    }
}

File: src/main/kotlin/it/frob/sleighidea/SleighParserDefinition.kt
// SPDX-License-Identifier: Apache-2.0

package it.frob.sleighidea

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.ParserDefinition.SpaceRequirements
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import it.frob.sleighidea.parser.SleighParser
import it.frob.sleighidea.psi.SleighTypes
import it.frob.sleighidea.psi.impl.SleighFileImpl

class SleighParserDefinition : ParserDefinition {

    override fun getWhitespaceTokens(): TokenSet = WHITE_SPACES

    override fun spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode): SpaceRequirements =
        SpaceRequirements.MAY

    override fun createLexer(project: Project): Lexer = SleighLexerAdapter()

    override fun createParser(project: Project): PsiParser = SleighParser()

    override fun getFileNodeType(): IFileElementType = FILE

    override fun getCommentTokens(): TokenSet = COMMENTS

    override fun getStringLiteralElements(): TokenSet = QSTRINGS

    override fun createElement(node: ASTNode): PsiElement = SleighTypes.Factory.createElement(node)

    override fun createFile(viewProvider: FileViewProvider): PsiFile = SleighFileImpl(viewProvider)

    companion object {
        val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
        val COMMENTS = TokenSet.create(SleighTypes.COMMENT)
        val QSTRINGS = TokenSet.create(SleighTypes.QUOTED_STRING)
        val FILE = IFileElementType(SleighLanguage.INSTANCE)
    }
}